components/autofill/src/sync/credit_card/incoming.rs

```diff
@@ -0,0 +1,296 @@
+    }
+
+    /// Returns a local record that has the same values as the given incoming record (with the exception
+    /// of the `guid` values which should differ) that will be used as a local duplicate record for
+    /// syncing.
+    fn get_local_dupe(
+        &self,
+        tx: &Transaction<'_>,
+        incoming: &Self::Record,
+    ) -> Result<Option<(SyncGuid, Self::Record)>> {
+        let sql = format!("
+            SELECT
+                {common_cols},
+                sync_change_counter
+            FROM credit_cards_data
+            WHERE
+                -- `guid <> :guid` is a pre-condition for this being called, but...
+                guid <> :guid
+                -- only non-synced records are candidates, which means can't already be in the mirror.
+                AND guid NOT IN (
+                    SELECT guid
+                    FROM credit_cards_mirror
+                )
+                -- and sql can check the field values.
+                AND cc_name == :cc_name
+                AND cc_number == :cc_number
+                AND cc_exp_month == :cc_exp_month
+                AND cc_exp_year == :cc_exp_year
+                AND cc_type == :cc_type", common_cols = CREDIT_CARD_COMMON_COLS);
+
+        let params = named_params! {
+            ":guid": incoming.guid,
+            ":cc_name": incoming.cc_name,
+            ":cc_number": incoming.cc_number,
+            ":cc_exp_month": incoming.cc_exp_month,
+            ":cc_exp_year": incoming.cc_exp_year,
+            ":cc_type": incoming.cc_type,
+        };
+
+        let result = tx.query_row_named(&sql, params, |row| {
+            Ok(Self::Record::from_row(&row).expect("wtf? '?' doesn't work :("))
+        });
+
+        match result {
+            Ok(r) => Ok(Some((incoming.guid.clone(), r))),
+            Err(e) => match e {
+                rusqlite::Error::QueryReturnedNoRows => Ok(None),
+                _ => Err(Error::SqlError(e)),
+            },
+        }
+    }
+
+    fn update_local_record(
+        &self,
+        tx: &Transaction<'_>,
+        new_record: Self::Record,
+        flag_as_changed: bool,
+    ) -> Result<()> {
+        update_internal_credit_card(tx, &new_record, flag_as_changed)?;
+        Ok(())
+    }
+
+    fn insert_local_record(&self, tx: &Transaction<'_>, new_record: Self::Record) -> Result<()> {
+        add_internal_credit_card(tx, &new_record)?;
+        Ok(())
+    }
+
+    /// Changes the guid of the local record for the given `old_guid` to the given `new_guid` used
+    /// for the `HasLocalDupe` incoming state, and mark the item as dirty.
+    fn change_local_guid(
+        &self,
+        tx: &Transaction<'_>,
+        old_guid: &SyncGuid,
+        new_guid: &SyncGuid,
+    ) -> Result<()> {
+        common_change_guid(tx, "credit_cards_data", old_guid, new_guid)
+    }
+
+    fn remove_record(&self, tx: &Transaction<'_>, guid: &SyncGuid) -> Result<()> {
+        common_remove_record(tx, "credit_cards_data", guid)
```

Comment:
nit: It might be worth making the table names `const`s that can be used here and exported and used in place of all the string literal table names. Not a showstopper though.